src/ring_theory/dedekind_domain/integer_unit.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2022 David Kurniadi Angdinata. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Kurniadi Angdinata
+-/
+
+import ring_theory.dedekind_domain.adic_valuation
+
+/-!
+# `S`-integers and `S`-units of fraction fields of Dedekind domains
+
+This file defines the ring of `S`-integers and the group of `S`-units of the field of fractions of a
+Dedekind domain, to be specialised to the case of a number field or a function field separately.
+
+TODO:
+* notation for `S`-integers and `S`-units.
+* proof that `S`-integers is the intersection of valuation rings.
+* proof that `S`-units is the kernel of a map to a product.
+* proof that `∅`-integers is the usual ring of integers.
+* finite generation of `S`-units and Dirichlet's `S`-unit theorem.
+
+## Main definitions
+
+* `is_dedekind_domain.S_integer`: the ring of `S`-integers.
+* `is_dedekind_domain.S_unit`: the group of `S`-units.
+
+## Main statements
+
+* `is_dedekind_domain.S_integer_unit`: the units of the `S`-integers are the `S`-units.
+
+## References
+
+* [D Marcus, *Number Fields*][marcus1977number]
+* [J W S Cassels, A Frölich, *Algebraic Number Theory*][cassels1967algebraic]
+* [J Neukirch, *Algebraic Number Theory*][Neukirch1992]
+
+## Tags
+
+S-integer, S-unit
+-/
+
+open_locale non_zero_divisors
+
+namespace is_dedekind_domain
+
+noncomputable theory
+
+universes u v
+
+variables {R : Type u} [comm_ring R] [is_domain R] [is_dedekind_domain R] (K : Type v) [field K]
+  [algebra R K] [is_fraction_ring R K] (S : set $ height_one_spectrum R)
+
+/-- The subring of `S`-integers of `K`. -/
+@[simps] def S_integer : subring K :=
```

Comment:
```suggestion
@[simps] def S_integers : subring K :=
```

Reply:
I don't know what others think of this: it's basically the question of whether we should have `x : S_integer` as in "let x be an S-integer", or `x : S_integers` as in "let x be an element of the ring of S-integers" (cf: `s : set X` for some type `X`)

Reply:
There was a short discussion on this on the Xena Discord server and @YaelDillies convinced me we should name it after the term whenever possible.

Reply:
I strongly encourage you to add the `s`. See e.g., https://leanprover-community.github.io/mathlib_docs/algebra/group/units.html#units

Reply:
I'm not a fan of the naming for `units`, but I'll add the `s` for consistency unless a global refactor happens.

Reply:
I suspect the only reason `units` is not called `unit` is because `unit` already means `punit.{0}`! Hence I don't think consistency stands here.

Reply:
I have reverted the naming to `integer` and `unit` based on the vote in Zulip.